### PR TITLE
[WIP] Ignore hidden or unreadable files and directories.

### PR DESCRIPTION
**Pull request by Adirelle**
This is a first implementation, related to https://github.com/anacrolix/dms/issues/44.

- [x] Test on Windows
- [ ] Test on Linux
- [x] Recursively check parents for hidden directories on Windows.
- [x] Added configuration options
- [x] Added command-line flags
- [x] Write automated tests ? (I don't know if it is possible)
- [x] Check file visibility before serving it.

Feedback is welcome.

### CODE REVIEW

File: dlna/dms/cds_unix.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+package dms
+
+import (
+	"strings"
+
+	"golang.org/x/sys/unix"
+)
+
+func isHiddenPath(path string) (bool, error) {
+	return strings.Contains(path, "/."), nil
```
Comment (anacrolix) [inline dlna/dms/cds_unix.go]: Will this work for hidden files immediately in the root?

File: dlna/dms/cds_unix.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+func isReadablePath(path string) (bool, error) {
+	err := unix.Access(path, unix.R_OK)
```
Comment (anacrolix) [inline dlna/dms/cds_unix.go]: I was thinking about this, it is technically ideal, but there are caveats that it's not guaranteed to be correct. I'm keen to try it anyway.

File: dlna/dms/cds_windows.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	"os"
+	"unicode/utf16"
+
+	"golang.org/x/sys/windows"
+)
+
+const hiddenAttributes = windows.FILE_ATTRIBUTE_HIDDEN | windows.FILE_ATTRIBUTE_SYSTEM
+
+func isHiddenPath(path string) (hidden bool, err error) {
+	attrs, err := windows.GetFileAttributes(toWindowsPath(path))
```
Comment (anacrolix) [inline dlna/dms/cds_windows.go]: Do you still want to treat dot files, for "unix-on-windows" style use of the filesystem or does this handle it?

**Review by anacrolix — APPROVED** [review]
Fantastic coding style. Are you able to test on Linux?

File: dlna/dms/cds_unix.go (hunk 1 above)
Comment (Adirelle) [reply dlna/dms/cds_unix.go]: Probably not. That's why I was thinking about unit testing.

File: dlna/dms/cds_unix.go (hunk 2 above)
Comment (Adirelle) [reply dlna/dms/cds_unix.go]: Where can I read these caveats ? Trying to guess the access by analyzing the permissions seems to be awfully complex.

File: dlna/dms/cds_windows.go (hunk 3 above)
Comment (Adirelle) [reply dlna/dms/cds_windows.go]: Nope. This only handles the Windows "hidden" and "system" attributes, which is Windows' idiomatic way to hide files.

File: dlna/dms/cds_unix.go (hunk 1 above)
Comment (anacrolix) [reply dlna/dms/cds_unix.go]: If we knew for certain that path started with /, then this is fine. Static analysis, or another method like always prepending a "/" would work.

File: dlna/dms/cds_windows.go (hunk 3 above)
Comment (Adirelle) [reply dlna/dms/cds_windows.go]: However, I think this will not check for parent status.

File: dlna/dms/cds_unix.go (hunk 2 above)
Comment (anacrolix) [reply dlna/dms/cds_unix.go]: Just in the man pages

**Comment by anacrolix** [comment]
I'm happy to merge this if you want. Or do you want to do more to it first?

**Comment by Adirelle** [comment]
I still have some uncheckd tasks.

**Comment by Adirelle** [comment]
Still have to test on Linux, but I am confident it should work. I think it is ok to merge it.

**Review by anacrolix — APPROVED** [review]

**Pull request merged**